src/coreclr/tools/aot/ILCompiler.Diagnostics/SymNgenWriterWrapper.cs

```diff
@@ -0,0 +1,145 @@
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+#nullable enable
+
+namespace Microsoft.DiaSymReader
+{
+    internal class SymNgenWriterWrapper : ISymNGenWriter2, IDisposable
+    {
+        private bool _isDisposed = false;
+        public IntPtr ISymNGenWriter2Inst { get; }
+
+        private SymNgenWriterWrapper(IntPtr writer2Inst)
+        {
+            ISymNGenWriter2Inst = writer2Inst;
+        }
+
+        public static SymNgenWriterWrapper? CreateIfSupported(IntPtr ptr)
+        {
+            var iid = ISymNGenWriter2.IID;
+            int hr = Marshal.QueryInterface(ptr, ref iid, out IntPtr ngenWriterInst);
+            if (hr != 0)
+            {
+                return null;
+            }
+
+            return new SymNgenWriterWrapper(ngenWriterInst);
+        }
+
+        ~SymNgenWriterWrapper()
+        {
+            DisposeInternal();
+        }
+
+        public void Dispose()
+        {
+            DisposeInternal();
+            GC.SuppressFinalize(this);
+        }
+
+        private void DisposeInternal()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+            Marshal.Release(ISymNGenWriter2Inst);
+            _isDisposed = true;
+        }
+
+        public unsafe void AddSymbol(string pSymbol, ushort iSection, ulong rva)
+        {
+            IntPtr strLocal = Marshal.StringToBSTR(pSymbol);
+            var inst = ISymNGenWriter2Inst;
+            var func = (delegate* unmanaged<IntPtr, IntPtr, ushort, ulong, int>)(*(*(void***)inst + 3 /* ISymNGenWriter2.AddSymbol slot */));
+            int hr = func(inst, strLocal, iSection, rva);
+            if (hr != 0)
+            {
+                Marshal.FreeBSTR(strLocal);
```

Comment:
Do we need to free the `strLocal` BSTR in successful cases as well, or is the ownership passed to the native object?

Reply:
Not 100% sure here (couldnt' find the code), but I think I need to always free it.